crates/criticalup-core/src/download_server_client.rs

```diff
@@ -113,13 +120,36 @@ impl DownloadServerClient {
     fn send(&self, builder: RequestBuilder) -> Result<Response, Error> {
         let req = builder.build().expect("failed to prepare the http request");
         let url = req.url().to_string();
-        let response = self
-            .client
-            .execute(req)
-            .map_err(|e| Error::DownloadServerError {
-                kind: DownloadServerError::Network(e),
-                url,
-            })?;
+
+        // Retry logic implemented here.
+        // This section implements a quick and dirty version of a retry a failed request for
+        // the max limit of `CLIENT_MAX_RETRIES` times with doubling the backoff each time
+        // starting with `CLIENT_RETRY_BACKOFF`.
+        //
+        // This logic will be superseded by changes to this code base when we move this to async.
+        let mut current_retries: u8 = 0;
+        let mut current_retry_backoff = CLIENT_RETRY_BACKOFF;
+
+        // This `try_clone()` will be `None` if request body is a stream.
+        let req_outer = req.try_clone().ok_or(Error::RequestCloningFailed)?;
+        let mut response_result = self.client.execute(req_outer);
+
+        while current_retries < CLIENT_MAX_RETRIES && response_result.is_err() {
+            // This try_clone()` will be `None` if request body is a stream.
```

Comment:
We create the req just above, and don't turn it into a stream at any point. We have note twice in the span of a few lines. Perhaps it's safe to remove it, since `req` should never be a stream, and users should never hit this error.